### PR TITLE
Merkle tree memoization

### PR DESCRIPTION
**Pull request by philandstuff**
PR raised so we have somewhere to capture comments

### CODE REVIEW

File: src/main/java/uk/gov/verifiablelog/merkletree/InMemory.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package uk.gov.verifiablelog.merkletree;
+import java.util.TreeMap;
+
+public class InMemory implements MemoizationStore {
+
+    private TreeMap<Integer, TreeMap<Integer, byte[]>> data;
```
Comment (philandstuff) [inline src/main/java/uk/gov/verifiablelog/merkletree/InMemory.java:6]: It'd be worth expressing why we went with `TreeMap<Integer, TreeMap<Integer, byte[]>>` rather than what I would have expected: `HashMap<Pair<Integer,Integer>,byte[]>`

File: src/main/java/uk/gov/verifiablelog/merkletree/InMemoryPowOfTwo.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    public InMemoryPowOfTwo() {
+        data = new TreeMap<>();
+    }
+
+
+    @Override
+    public void put(Integer start, Integer size, byte[] value) {
+        if (((start == 0) || (Integer.bitCount(start) == 1)) &&
+                ((size == 0) || (Integer.bitCount(size) == 1))) {
```
Comment (philandstuff) [inline src/main/java/uk/gov/verifiablelog/merkletree/InMemoryPowOfTwo.java:18]: maybe could use `<= 1` rather than `== 1` to include the zero case? is that clearer?

also is it worth extracting a method with an intention-revealing name such as `isZeroOrPowerOfTwo`?

Comment (philandstuff) [inline src/main/java/uk/gov/verifiablelog/merkletree/InMemoryPowOfTwo.java:18]: also we should (at some point) test this class. shouldn't block merging this PR though.

File: src/main/java/uk/gov/verifiablelog/merkletree/MerkleTree.java [hunk 3]
```diff
@@ -11,83 +11,107 @@
...
-            subtreePath.add(subtreeHash(k, subtreeDAOFunction));
+            List<byte[]> subtreePath = subtreePathAtSnapshot(leafIndex - k, start + k, snapshotSize - k);
+            subtreePath.add(realSubtreeHash(start, k));
             return subtreePath;
         }
     }
 
     // hash of subtree of given size
-    private byte[] subtreeHash(int size, Function<Integer, byte[]> leafDAOFunction) {
+    private byte[] realSubtreeHash(int n, int size) {
```
Comment (philandstuff) [inline src/main/java/uk/gov/verifiablelog/merkletree/MerkleTree.java:86]: from @michaelabenyohai : we should use `start` not `n` here for consistency

File: src/test/java/uk/gov/verifiablelog/merkletree/MerkleTreeTests.java [hunk 4]
```diff
@@ -29,114 +31,157 @@
...
     @Before
-    public void setup() throws NoSuchAlgorithmException {
-        leafValues = new ArrayList<>();
-        merkleTree = new MerkleTree(MessageDigest.getInstance("SHA-256"), leafValues::get, leafValues::size);
+    public void beforeEach() throws NoSuchAlgorithmException {
+        MemoizationStore inMemory = new InMemory();
+        MemoizationStore inMemoryPowOfTwo = new InMemoryPowOfTwo();
+        merkleTreeTestUnits = new ArrayList<>(Arrays.asList(
+                makeMerkleTreeTestUnit(null),
+                makeMerkleTreeTestUnit(inMemory),
```
Comment (philandstuff) [inline src/test/java/uk/gov/verifiablelog/merkletree/MerkleTreeTests.java:45]: we don't need to test a use case where we have a merkle tree with an empty leaf list but a nonempty MemoizationStore.

**Comment by philandstuff** [comment]
merging, but we still need to address some of the comments
